fix: Implement TopNavigation breakpoints in CSS

TrevorBurnham opened the pull request:
### Description

This PR switches from JavaScript-based breakpoints to CSS container queries for TopNavigation, making its appearance consistent when rendered with SSR. I've added Sass mixins for this purpose that can be reused across other components to replace the `useBreakpoints` hook.

I've opted to use container queries rather than media queries for backward compatibility's sake, since `useBreakpoints` checks container size rather than viewport size. However, this introduces a complication: To use container queries, we need to add `container-type: inline-size` to the TopNavigation element's styles.

I noticed that adding `container-type: inline-size` causes overflowing content to get clipped in Safari due to a longstanding bug in that browser. [**Update 4/5:** After updating to Safari 18.4 on my Mac, I no longer see the bug, so it looks like Apple just fixed this.] To address that, I've enabled `expandToViewport` to render utility dropdowns in a portal. I've also added `position: relative` and `z-index: 1` to the `TopNavigation` styles to give it a stacking context, which appears to help with the issue. However, there is still a possibility that users who are rendering their own dropdowns in the `TopNavigation` (e.g. by using an `Autosuggest` searchbox) without `expandToViewport` may see issues in older versions of Safari where the dropdown is rendered behind other content with a higher `z-index`. I would suggest calling this out in the release notes and recommending the use of `expandToViewport` for dropdowns in `TopNavigation`.

Once this change is merged, I'd love to start applying this same change to other components that use `useBreakpoints`. Container queries are [supported in all modern browsers](https://caniuse.com/css-container-queries).

Fixes #3337 

### How has this been tested?

I've run this change locally, eyeballed it at different breakpoints, and clicked various controls to ensure that they're visible. I'm open to ideas for testing the breakpoint behavior more thoroughly.

<details>
   <summary>Review checklist</summary>

_The following items are to be evaluated by the author(s) and the reviewer(s)._

#### Correctness

- _Changes include appropriate documentation updates._
- _Changes are backward-compatible if not indicated, see [`CONTRIBUTING.md`](https://github.com/cloudscape-design/components/blob/main/CONTRIBUTING.md#public-apis)._
- _Changes do not include unsupported browser features, see [`CONTRIBUTING.md`](https://github.com/cloudscape-design/components/blob/main/CONTRIBUTING.md#browsers-support)._
- _Changes were manually tested for accessibility, see [accessibility guidelines](https://cloudscape.design/foundation/core-principles/accessibility/)._

#### Security

- _If the code handles URLs: all URLs are validated through [the `checkSafeUrl` function](https://github.com/cloudscape-design/components/blob/main/src/internal/utils/check-safe-url.ts)._

#### Testing

- _Changes are covered with new/existing unit tests?_
- _Changes are covered with new/existing integration tests?_
</details>

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

pan-kot commented:
Hello @TrevorBurnham,

Thanks for contributing the fix!

Our team is taking a look and running internal screenshot tests and manual tests.

Could you please fix the failing unit tests? P.S. the "Measure bundle size" action is expected to fail, please ignore it.

TrevorBurnham commented:
@pan-kot Sure thing. I've pushed a fix for the unit tests: [51e8d78](https://github.com/cloudscape-design/components/pull/3371/commits/51e8d7828897b623d90c717fbd0e0b391e454a39)

TrevorBurnham commented:
Thanks! Can you give me instructions for viewing the component before/after? What I've been doing is running `npm run build` and then running a project against the built components directory, but some of the styles seem to be lost in that process (possibly because the "visual refresh" flag doesn't work).

pan-kot commented:
> Thanks! Can you give me instructions for viewing the component before/after? What I've been doing is running `npm run build` and then running a project against the built components directory, but some of the styles seem to be lost in that process (possibly because the "visual refresh" flag doesn't work).

The npm run build should be fine: you can build the mainline and then your commit, and compare the differences. The `./pages/lib` folder should contain the static assets for the dev pages.

TrevorBurnham commented:
I've updated the PR branch to fix the visual inconsistencies and improve code coverage.

One call-out: Previously `TopNavigation` passed a prop called `offsetRight` to the `ButtonTrigger` for utility menu dropdowns, then added styles based on the `.offset-right-*` class. Since the `offsetRight` value depended on the breakpoint in JavaScript, that approach doesn't work; there needs to be a CSS selector that can be used in the `TopNavigation` styles to reference the `ButtonTrigger` element. So I added a prop called `internalClass` to allow a classname to be passed through. Let me know if there's a better way to do this.

pan-kot commented on src/internal/components/menu-dropdown/index.tsx:
```diff
@@ -20,6 +20,7 @@ export const ButtonTrigger = React.forwardRef(
   (
     {
       testUtilsClass,
+      internalClass,
```
We normally call this `className`, when it is applied to the top element. I would also add a comment, explicitly documenting the use case.

TrevorBurnham replied:
Done, thanks!

pan-kot commented on src/test-utils/dom/top-navigation/index.ts:
```diff
@@ -101,10 +101,10 @@ export class TopNavigationMenuDropdownWrapper extends ButtonDropdownWrapper {
 
   findTitle(): ElementWrapper | null {
-    return this.findByClassName(buttonDropdownStyles.title);
+    return createWrapper().findComponent(`.${buttonDropdownStyles.title}`, ElementWrapper);
```
What is the reason of this change?

TrevorBurnham replied:
Without this change, the test

```
TopNavigation Utility part › Menu dropdown › has a title in the dropdown
```

fails to find the title element, due to the menu dropdown having `expandToViewport`.

pan-kot commented on src/test-utils/dom/button-dropdown/index.ts:
```diff
@@ -40,20 +40,30 @@ export default class ButtonDropdownWrapper extends ComponentWrapper {
    * Finds an item in the open dropdown by item id. Returns null if there is no open dropdown.
    *
    * This utility does not open the dropdown. To find dropdown items, call `openDropdown()` first.
+   *
```
There are many button dropdown tests that use expandToViewport (see example: https://github.com/cloudscape-design/components/blob/main/src/button-dropdown/__tests__/button-dropdown-highlighted.test.tsx#L24), but the expandToViewport option wasn't needed. I think that is because the `findOpenDropdown()` method searches from the root, not component root - assuming there is at most one open dropdown. Why do we need the expandToViewport option now?

TrevorBurnham replied:
Good call-out. This change wasn't needed.

pan-kot commented on src/internal/styles/foundation/breakpoints.scss:
```diff
@@ -39,3 +51,27 @@ $_largest_breakpoint: $breakpoint-x-large;
   }
 }
+
+// Container query for widths greater than the given breakpoint.
```
Nice 👍

TrevorBurnham commented on src/top-navigation/styles.scss:
```diff
@@ -22,18 +24,52 @@
 
-  &.narrow > .padding-box {
-    block-size: awsui.$space-xxxl;
+  // Narrow container (width < 465px)
```
Done, thanks.

TrevorBurnham commented:
I've rebased this branch and made the following changes:

1. Added `position: relative` and `z-index: 1` to the `TopNavigation` to work around a Safari bug where overflow content from containers with `container-type: inline-size` can be incorrectly rendered behind other elements, causing clipping. I've updated the PR description with more details about this bug, which I think is the most serious potential issue with this change.
2. Renamed `internalClassName` to just `className` and added an explanatory comment re: feedback.
3. Reduced unnecessary test changes and comments re: feedback.

TrevorBurnham committed:
fix: Implement TopNavigation breakpoints in CSS

This commit switches from JavaScript-based breakpoints to CSS container
queries for TopNavigation, making its appearance consistent when
rendered with SSR.

To support container queries, we need `container-type: inline-size;` on
the TopNavigation root element. This causes overflowing content to get
clipped (at least in Safari), so I've enabled `expandToViewport` to
render utility dropdowns in a portal. Note that OverflowMenu doesn't
need this treatment because it expands the height of TopNavigation
rather than overflowing.

Fixes #3337

TrevorBurnham commented:
@pan-kot I've rebased and updated this branch.

TrevorBurnham committed:
Updates in response to feedback

TrevorBurnham commented on src/internal/components/menu-dropdown/interfaces.ts:
```diff
@@ -19,12 +20,14 @@ export interface ButtonTriggerProps {
 
-export interface MenuDropdownProps extends InternalButtonDropdownProps {
+export interface MenuDropdownProps extends Omit<InternalButtonDropdownProps, 'items'> {
+  items: InternalButtonDropdownProps['items'];
```
Good call-out! I've removed this change.

TrevorBurnham committed:
Merge branch 'main' into topnavigation-breakpoints-3337
